pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,2106 @@
+        self.rhoinf = flowinfo[0]
+        self.uinf = flowinfo[1:self.ndims+1]
+        self.pinf = flowinfo[-1]
+
+    def acoustic_psoln(self, *args):
+        pass
+    def update_usoln(self, *args):
+        pass
+
+class MonopoleSrc(PointAcousticSrc):
+    def __init__(self, name, srclocs, flowinfo, ampl, srcfreq, ntime, nperiods,
+                 gamma):
+        super().__init__(name, srclocs, flowinfo, ampl, srcfreq, ntime,
+                         nperiods, gamma)
+        self.gamma = gamma
+        self.co = np.sqrt(self.gamma*self.pinf/self.rhoinf)
+        self.ampl = ampl
+        self.srcfreq = srcfreq
+        self.nperiods = nperiods
+        self.nt = ntime
+        self.omega = 2.*np.pi*self.srcfreq
+        self.tperiod = 2.*np.pi/self.omega
+        self.dt = self.nperiods*self.tperiod/self.nt
+        self.freq = np.fft.rfftfreq(self.nt,self.dt)
+        self.nfreq = len(self.freq)
+        
+    def acoustic_psoln(self, xyz_ob):
+        nobserv = xyz_ob.shape[0]
+        ptime = np.empty((self.nt, nobserv))
+        pfft = np.empty((nobserv, self.nfreq), dtype=np.complex64)
+
+        tarr = np.arange(0, self.nt)*self.dt
+        for i, ti in enumerate(tarr):
+            ptime[i] = self._comput_exact_psoln(xyz_ob, ti, ptime[i])
+
+        ptime = np.moveaxis(ptime, 0, 1)
+        pfft = realfft(ptime)
+
+        return self.freq, pfft
+
+    def _comput_exact_psoln(self, xyz_ob, tcurr, psoln):
+        phy, _, mdotdphy = self._comput_vel_potentials(xyz_ob, tcurr)
+        psoln = - np.real(self.rhoinf*(1j*self.omega*phy + self.co*mdotdphy))
+        return psoln
+
+    def update_usoln(self, xyz_ob, tcurr, usoln):
+        co = self.co
+        phy, dphy, mdotdphy = self._comput_vel_potentials(xyz_ob, tcurr)
+        # computing the flow quantities
+        p = - np.real(self.rhoinf*(1j*self.omega*phy + co*mdotdphy))
+        usoln[-1] = p + self.pinf
+        usoln[0] = p/(co*co) + self.rhoinf  #rho
+        usoln[1: self.ndims+1] = np.array([ui + uo for ui, uo in
+                                    zip(np.real(dphy), self.uinf)]) #u
+
+    def _comput_vel_potentials(self, xyz_ob, tcurr):
+        Mo = self.uinf/self.co
+        kwv = self.omega/self.co
+        xyz_src = self.srclocs
+        mR, mRs, nR, nRs = FwhFreqDomainSolver.compute_distance_vars(xyz_src,
+                                                                    xyz_ob, Mo)
+        # phy_potential of the source and its derivatives
+        amp_4piRs = self.ampl/(4.*np.pi*mRs)
+        phy = amp_4piRs*np.exp(1j*(self.omega*tcurr - kwv*mR)) 
+        invRs = 1./mRs
+        kk = 1j*kwv*nR + np.einsum('i,ij->ij', invRs, nRs)
+        dphy = - np.einsum('i,ij->ji', phy, kk) 
+        mdotdphy = sum((Mo*dphy.T).T)
+        return phy, dphy, mdotdphy
+
+class DipoleSrc(PointAcousticSrc):
+    def __init__(self, name, srclocs, flowinfo, *srcdata):
+        super().__init__(name, srclocs, flowinfo, *srcdata)
+
+class VTUSurfWriter(object):
```

Comment:
I'm guessing this is for debugging purposes?

Reply:
Yes